tests/unit/libexpr/trivial.cc

```diff
@@ -177,6 +177,48 @@ namespace nix {
             )
     );
 
+#define X_EXPAND_IF0(k, v) k "." v
+#define X_EXPAND_IF1(k, v) k " = { " v " };"
+#define X4(w, x, y, z) \
+    TEST_F(TrivialExpressionTest, nestedAttrsetMerge##w##x##y##z) { \
+        auto v = eval("{ a.b = { c = 1; d = 2; }; } == { " \
+                X_EXPAND_IF##w("a", X_EXPAND_IF##x("b","c = 1;")) " " \
+                X_EXPAND_IF##y("a", X_EXPAND_IF##z("b","d = 2;")) " }"); \
+        ASSERT_THAT(v, IsTrue()); \
+    }; \
+    TEST_F(TrivialExpressionTest, nestedAttrsetMergeDup##w##x##y##z) { \
+        ASSERT_THROW(eval("{ " \
+                X_EXPAND_IF##w("a", X_EXPAND_IF##x("b","c = 1;")) " " \
+                X_EXPAND_IF##y("a", X_EXPAND_IF##z("b","c = 2;")) " }"), Error); \
+    }; \
+    TEST_F(TrivialExpressionTest, nestedAttrsetMergeLet##w##x##y##z) { \
+        auto v = eval("{ b = { c = 1; d = 2; }; } == (let " \
+                X_EXPAND_IF##w("a", X_EXPAND_IF##x("b","c = 1;")) " " \
+                X_EXPAND_IF##y("a", X_EXPAND_IF##z("b","d = 2;")) " in a)"); \
+        ASSERT_THAT(v, IsTrue()); \
+    };
+#define X3(...) X4(__VA_ARGS__, 0) X4(__VA_ARGS__, 1)
+#define X2(...) X3(__VA_ARGS__, 0) X3(__VA_ARGS__, 1)
+#define X1(...) X2(__VA_ARGS__, 0) X2(__VA_ARGS__, 1)
+    X1(0) X1(1)
+#undef X_EXPAND_IF0
+#undef X_EXPAND_IF1
+#undef X1
+#undef X2
+#undef X3
+#undef X4
+
+    // This is for backwards compatibility, not because we like it.
+    // See https://github.com/NixOS/nix/issues/9020.
+    TEST_F(TrivialExpressionTest, regrettableRecAttrsetMerge) {
+        auto v = eval("{ a = rec { b = c + 1; d = 2; }; a.c = d + 3; }.a.b");
+        ASSERT_THAT(v, IsIntEq(6));
+    }
+
+    TEST_F(TrivialExpressionTest, attrsetMergeDropsLaterRec) {
+        ASSERT_THROW(eval("{ a.b = 1; a = rec { c = d + 2; d = 3; }; }.c"), Error);
+    }
```

Comment:
Maybe move these to tests/functional/lang? That's the preferred place for language tests.